apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClientImpl.cs

```diff
@@ -0,0 +1,204 @@
+    /// <param name="shutdown">Function to call on this <see cref="SubscriberClientImpl"/> shutdown.</param>
+    public SubscriberClientImpl(SubscriptionName subscriptionName, IEnumerable<SubscriberServiceApiClient> clients, Settings settings, Func<Task> shutdown)
+        : this(subscriptionName, clients, settings, shutdown, TaskHelper.Default) { }
+
+    internal SubscriberClientImpl(SubscriptionName subscriptionName, IEnumerable<SubscriberServiceApiClient> clients, Settings settings, Func<Task> shutdown, TaskHelper taskHelper)
+    {
+        SubscriptionName = GaxPreconditions.CheckNotNull(subscriptionName, nameof(subscriptionName));
+        GaxPreconditions.CheckNotNull(clients, nameof(clients));
+        _clients = clients.ToArray();
+        GaxPreconditions.CheckArgument(_clients.Length > 0, nameof(clients), "Must contain at least one client");
+        GaxPreconditions.CheckArgument(_clients.All(x => x != null), nameof(clients), "All elements must be non-null");
+        GaxPreconditions.CheckNotNull(settings, nameof(settings));
+        settings.Validate();
+        // These values are validated in Settings.Validate() above, so no need to re-validate here.
+        _modifyDeadlineSeconds = (int)((settings.AckDeadline ?? DefaultAckDeadline).TotalSeconds);
+        var autoExtendInterval = TimeSpan.FromSeconds(_modifyDeadlineSeconds) - (settings.AckExtensionWindow ?? DefaultAckExtensionWindow);
+        var autoExtendIntervalForExactlyOnceDelivery = settings.AckExtensionWindow ?? MinimumAckExtensionWindowForExactlyOnceDelivery;
+        // For exactly once delivery subscription, minimum ack extension window value should be the default value of 60 seconds or the user provided value.
+        // Ensure the duration between lease extensions is at least MinimumLeaseExtensionDelay (5 seconds).
+        // The minimum allowable lease duration is 10 seconds, so this will always be reasonable.
+        _autoExtendInterval = TimeSpan.FromTicks(Math.Max(autoExtendInterval.Ticks, MinimumLeaseExtensionDelay.Ticks));
+        _autoExtendIntervalForExactlyOnceDelivery = TimeSpan.FromTicks(Math.Max(autoExtendIntervalForExactlyOnceDelivery.Ticks, MinimumLeaseExtensionDelay.Ticks));
+        _maxExtensionDuration = settings.MaxTotalAckExtension ?? DefaultMaxTotalAckExtension;
+        _shutdown = shutdown;
+        _scheduler = settings.Scheduler ?? SystemScheduler.Instance;
+        _clock = settings.Clock ?? SystemClock.Instance;
+        _taskHelper = GaxPreconditions.CheckNotNull(taskHelper, nameof(taskHelper));
+        _flowControlSettings = settings.FlowControlSettings ?? DefaultFlowControlSettings;
+        _useLegacyFlowControl = settings.UseLegacyFlowControl;
+        _maxAckExtendQueue = (int)Math.Min(_flowControlSettings.MaxOutstandingElementCount ?? long.MaxValue, 20_000);
+        _disposeTimeout = settings.DisposeTimeout ?? DefaultDisposeTimeout;
+    }
+
+    private readonly object _lock = new object();
```

Comment:
Please could you move all these fields and properties to before the constructor while we're at it?

(This is very old code, before we'd adopted some of our current conventions.)

Reply:
(I didn't spot whether the same is true in other files, but it would be good to just go through them.)